Use mbedtls MPI (bignum) routines from LTC

jenswi-linaro opened the pull request:
This replaces libmpa usage in LTC via config option `CFG_MBEDTLS_MPI=y`. With all the commits in this PR there's quite some performance gain. Ironically also libmpa is improved a bit, but perhaps not when it comes to key generation. Here's some figures when testing on Hikey:

Four test cases from xtest has been selected for performance measurment
- 4006 -- asymmetric signing and verification
- 4007 -- key generation
- 4008 -- key derivation
- 4009 -- ECDH key derivation

The figures has been collected with the following script:
``` bash
for a in 4006 4007 4008 4009 ; do \
echo -n $a " " >> time.txt ;\
time -o time.txt.tmp xtest -l 15 $a || break ;\
grep real time.txt.tmp >> time.txt
done
cat time.txt
```

```
MPA eb32eb0b4961e121a "tee_time_arm_cntpct.c: remove mpa.h dependency"
4006  real      4m 37.37s
4007  real      1m 11.10s
4008  real      0m 0.13s
4009  real      2m 46.97s


MPA "core: mbedtls: configure bignum word size"
4006  real      3m 11.61s
4007  real      1m 17.04s
4008  real      0m 0.13s
4009  real      1m 46.97s

MPI "core: mbedtls: configure bignum word size"
4006  real      1m 36.04s
4007  real      0m 12.32s
4008  real      0m 0.12s
4009  real      1m 2.04s
```

Especially the key generation case (4007) using MPI instead of MPA stands out.

jbech-linaro commented:
Sweet, that looks pretty good (performance)!

jforissier commented on core/core.mk:
```diff
@@ -83,13 +83,24 @@ libname = utils
+CFG_CRYPTOLIB_NAME ?= tomcrypt
+CFG_CRYPTOLIB_DIR ?= core/lib/libtomcrypt
+
+CFG_MBEDTLS_MPI ?= y
```
I think the symbol name should be more self-explanatory, here it is not clear it is for the TEE core only. `CFG_TEE_CORE_BIGNUM_MBEDTLS[_MPI]" perhaps?
Similarly, I'd rather the subject be "core: switch to mbed TLS for bignum".

jenswi-linaro replied:
I was planning letting user TAs switch to use the MPI also using this config variable too. I guess it makes sense to be able to select bignum implementation independently between core and user TAs. In that case I'd prefer using `CFG_CORE_MBEDTLS_MPI`. The rational behind that is that it's something that is enabled in mbedtls when compiled for Core.

We could add another flag for LTC to use MPI instead of MPA or throw all that abbreviations into one big name but that seems a bit clumsy.

This is a transition, once we've stopped using libmpa completely and the dust has settled a bit we'll just remove libmpa and I suppose `CFG_CORE_MBEDTLS_MPI` will look quite natural.

I'll update the subject.

jforissier replied:
> I guess it makes sense to be able to select bignum implementation independently between core and user TAs. 

Indeed it was my main point.

> In that case I'd prefer using `CFG_CORE_MBEDTLS_MPI`.

Fine with me.

jforissier commented on lib/libutils/ext/mempool.c:
```diff
@@ -61,26 +62,28 @@ struct mempool {
 static void get_pool(struct mempool *pool __maybe_unused)
 {
 #if defined(__KERNEL__)
-	mutex_lock(&pool->mu);
```
Is `mutex_lock()`/`mutex_unlock()` much slower than using `refcount_*` in this case?

jenswi-linaro replied:
Yes, I think this is why libmpa was improved quite a bit too. Before I started experimenting with this the MPI performance wasn't that impressive.

jforissier replied:
Hmmm, OK, then perhaps there is something wrong with the mutex fast path? Or perhaps it's the debug code? Anyway, the change is certainly fine by itself.

jenswi-linaro replied:
The problem with the fast path is that it does too much, but still it can never become as good as a single `refcount_inc()`.

jforissier replied:
Sure.

etienne-lms commented on core/lib/libtomcrypt/src/mpi_desc.c:
```diff
@@ -0,0 +1,738 @@
+#include <util.h>
+#include <mempool.h>
+
+#define MPI_MEMPOOL_SIZE	(42 * 1024)
```
Why 42kB? I think libpma is aournd 12kB for 50 numbers of 4096bits.
I agree an oversized value won't hurt pager constraints. 
Yet, maybe a comment stating why this size could be nice.

jenswi-linaro replied:
The libmpa size is 52200 bytes
I'll add a comment on the size.

etienne-lms commented on lib/libmbedtls/mbedtls/include/mbedtls/bignum.h:
```diff
@@ -747,6 +751,41 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int dh_flag,
+
+/**
+ * \brief          Montgomery multiplication: A = A * B * R^-1 mod N
+ * \A              parameter and result
```
All other descriptions in this file uses uppercase for params. Should fix.

jenswi-linaro replied:
I'll fix

etienne-lms commented on lib/libmbedtls/mbedtls/library/bignum.c:
```diff
@@ -100,7 +116,10 @@ void mbedtls_mpi_free( mbedtls_mpi *X )
     {
         mbedtls_mpi_zeroize( X->p, X->n );
-        mbedtls_free( X->p );
+        if (X->use_mempool)
```
nitpicking on mbedtls coding style, i saw that there should be space around parentheses whatever the parenthesis is used for.

jenswi-linaro replied:
I'll fix

jenswi-linaro commented on lib/libutils/ext/mempool.c:
```diff
@@ -167,6 +170,13 @@ void *mempool_alloc(struct mempool *pool, size_t size)
+	if (pool->last_offset > pool->max_last_offset) {
+		pool->max_last_offset = pool->last_offset;
+		DMSG("max memory usage increased to %zu",
+		     (size_t)pool->max_last_offset);
```
The cast is needed as `max_last_offset` is a `ssize_t`. One would have expected `%zd` to work with that but it seems `ssize_t` need to be defined differently then.
I'll fix the upper case.

jenswi-linaro commented on core/lib/libtomcrypt/src/mpi_desc.c:
```diff
@@ -0,0 +1,738 @@
+#include <mm/core_mmu.h>
+
+/* allocate pageable_zi vmem for mpa scratch memory pool */
+static struct mempool *get_mpa_scratch_memory_pool(void)
```
I'll fix

jforissier commented:
You need this to fix the Shippable build errror:
```diff
diff --git a/lib/libmbedtls/sub.mk b/lib/libmbedtls/sub.mk
index 06c38baa8..01897404e 100644
--- a/lib/libmbedtls/sub.mk
+++ b/lib/libmbedtls/sub.mk
@@ -87,4 +87,6 @@ srcs-$(sm-$(ta-target)) += $(addprefix mbedtls/library/, $(SRCS_TLS))
 
 cflags-lib-y += -Wno-redundant-decls
 cflags-lib-y += -Wno-switch-default
+ifneq ($(sm),core) # User-mode
 cflags-lib-$(CFG_ULIBS_GPROF) += -pg
+endif
```

jenswi-linaro commented:
Applied the patch. Yes, `-pg` should probably be applied in a central manner.

jforissier commented:
For "[review] libmbedtls: configure and compile mbedtls":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

I guess it should go to the import branch, too.

jenswi-linaro commented:
If it's needed on the import branch it should probably be added via another merge from master as this isn't a change to the imported source code.

jforissier commented:
OK, makes sense.

jenswi-linaro commented:
Squashed and rebased the old tip (96622dbadf0be014a87e82324f2c9e8787f43ae8) and cherry-picked the recent `import/mbedtls-2.6.1` additions instead of the temporary mbedtls commits in this PR. Added Jerome's R-B to "libmbedtls: configure mbedTLS for different modes".

jforissier commented:
- For the three "mempool: *" commits as well as for "LTC: drop LTC_NO_ASM and configure" and "core: mbedtls: configure bignum word size":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "core: switch to mbedtls for bignum":
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Tags applied
